data-prepper-plugins/kinesis-source/src/main/java/org/opensearch/dataprepper/plugins/kinesis/source/processor/KinesisRecordProcessor.java

```diff
@@ -0,0 +1,173 @@
+        this.acknowledgementSetCallbackCounter = pluginMetrics.counterWithTags(ACKNOWLEDGEMENT_SET_CALLBACK_METRIC_NAME, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName());
+        this.recordProcessingErrors = pluginMetrics.counterWithTags(KINESIS_RECORD_PROCESSING_ERRORS, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName());
+        this.checkpointFailures = pluginMetrics.counterWithTags(KINESIS_CHECKPOINT_FAILURES, KINESIS_STREAM_TAG_KEY, streamIdentifier.streamName());
+        this.checkpointIntervalMilliSeconds = kinesisStreamConfig.getCheckPointIntervalInMilliseconds();
+        this.buffer = buffer;
+    }
+
+    private KinesisStreamConfig getStreamConfig(final KinesisSourceConfig kinesisSourceConfig) {
+        return kinesisSourceConfig.getStreams().stream().filter(streamConfig -> streamConfig.getName().equals(streamIdentifier.streamName())).findAny().get();
+    }
+
+    @Override
+    public void initialize(InitializationInput initializationInput) {
+        // Called once when the processor is initialized.
+        kinesisShardId = initializationInput.shardId();
+        LOG.info("Initialize Processor for shard: " + kinesisShardId);
+        lastCheckpointTimeInMillis = System.currentTimeMillis();
+    }
+
+    private AcknowledgementSet createAcknowledgmentSet(final ProcessRecordsInput processRecordsInput) {
+        return acknowledgementSetManager.create((result) -> {
+            acknowledgementSetCallbackCounter.increment();
+            if (result) {
+                LOG.info("acknowledgements received");
+                checkpoint(processRecordsInput.checkpointer());
+            } else {
+                LOG.info("acknowledgements received with false");
+            }
+
+        }, ACKNOWLEDGEMENT_SET_TIMEOUT);
+    }
+
+    @Override
+    public void processRecords(ProcessRecordsInput processRecordsInput) {
+        List<Record<Event>> records = new ArrayList<>();
+
+        try {
+            Optional<AcknowledgementSet> acknowledgementSetOpt = Optional.empty();
+            boolean acknowledgementsEnabled = kinesisSourceConfig.isAcknowledgments();
+            if (acknowledgementsEnabled) {
+                acknowledgementSetOpt = Optional.of(createAcknowledgmentSet(processRecordsInput));
+            }
+
+            for (KinesisClientRecord record : processRecordsInput.records()) {
+                processRecord(record, records::add);
+            }
+
+            acknowledgementSetOpt.ifPresent(acknowledgementSet -> records.forEach(record -> acknowledgementSet.add(record.getData())));
+
+            buffer.writeAll(records, bufferTimeoutMillis);
+
+            acknowledgementSetOpt.ifPresent(AcknowledgementSet::complete);
+
+            // Checkpoint for shard
+            if (kinesisStreamConfig.isEnableCheckPoint() && System.currentTimeMillis() - lastCheckpointTimeInMillis > checkpointIntervalMilliSeconds) {
+                LOG.info("Regular checkpointing for shard " + kinesisShardId);
```

Comment:
```suggestion
                LOG.debug("Regular checkpointing for shard {}", kinesisShardId);
```

Use Sl4fj string interpolation.

Also, this is probably more of a debug log.

Reply:
Thanks @dlvenable. I have fixed this.

Reply:
Updating the checkpoint synchronously is a pattern that we should move away from, as it can lead to this issue (https://github.com/opensearch-project/data-prepper/issues/4422). You can create a separate thread to do this checkpointing. To handle it for acknowledgments, you can add a progress check (example here: https://github.com/opensearch-project/data-prepper/pull/4918). Unless KCL handles the ownership aspect and the checkpointing is just to show how far it has gone in a shard